packages/sanity/src/core/form/inputs/arrays/ArrayOfObjectsInput/InsertMenuMenuItems.tsx

```diff
@@ -0,0 +1,118 @@
+import {InsertAboveIcon, InsertBelowIcon} from '@sanity/icons'
+import {type InsertMenuOptions} from '@sanity/insert-menu'
+import {type SchemaType} from '@sanity/types'
+import {useCallback, useMemo} from 'react'
+import {useTranslation} from 'sanity'
+
+import {MenuItem} from '../../../../../ui-components'
+import {useInsertMenuPopover} from './InsertMenuPopover'
+
+/**
+ * @internal
+ */
+type InsertMenuItemsProps = {
+  insertMenuOptions?: InsertMenuOptions
+  onInsert: (pos: 'before' | 'after', type: SchemaType) => void
+  referenceElement: HTMLElement | null
+  schemaTypes?: SchemaType[]
+}
+
+/**
+ * @internal
+ */
+export function useInsertMenuMenuItems(props: InsertMenuItemsProps) {
+  const {t} = useTranslation()
+  const {onInsert, schemaTypes: types} = props
+  const insertBefore = useInsertMenuPopover({
+    insertMenuProps: {
+      ...props.insertMenuOptions,
+      schemaTypes: props.schemaTypes ?? [],
+      onSelect: (insertType) => {
+        props.onInsert('before', insertType)
+      },
+    },
+    popoverProps: {
+      referenceElement: props.referenceElement,
+      placement: 'top-end',
```

Comment:
do you think it would be better to have this be consistent with `insertAfter`? I personally think setting them both to prefer `'bottom-end'` feels better.

This video is the current experience.


https://github.com/sanity-io/sanity/assets/10551026/c6e01f5c-332c-481c-a958-306080b1a350

Reply:
The decision to alter between `top-end` and `bottom-end` is primarily made by @mariuslundgard. The idea is that the placement gives a little more context in terms of where the item is inserted (above or below).